File: dnas/sensemaker/zomes/integrity/sensemaker/src/resource_type.rs
use hdi::prelude::*;

use crate::applet::ConfigResourceDef;

#[hdk_entry_helper]
#[derive(Clone)]
pub struct ResourceDef {
    pub name: String,
    pub base_types: Vec<AppEntryDef>,
    pub dimension_ehs: Vec<EntryHash>,
}

impl TryFrom<ConfigResourceDef> for ResourceDef {
    type Error = WasmError;
    fn try_from(value: ConfigResourceDef) -> Result<Self, Self::Error> {
        let dimension_ehs = value
            .dimensions
            .into_iter()
            .map(|dimension| hash_entry(dimension))
            .collect::<ExternResult<Vec<EntryHash>>>()?;
        let resource_type = ResourceDef {
            name: value.name,
            base_types: value.base_types,
            dimension_ehs,
        };
        Ok(resource_type)
    }
}

File: types/src/resourceType.ts
import { AppEntryDef, EntryHash } from "@holochain/client";
import { Dimension } from "./dimension";

interface CoreResourceDef {
    name: string,
    base_types: Array<AppEntryDef>,
}
export type ResourceDef = CoreResourceDef & {
    dimension_ehs: Array<EntryHash>,
}

export type ConfigResourceDef = CoreResourceDef & {
    dimensions: Array<Dimension>,
}